service/src/main/java/bio/terra/profile/db/ProfileChangeLogDao.java

```diff
@@ -0,0 +1,154 @@
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class ProfileChangeLogDao {
+  private static final Logger logger = LoggerFactory.getLogger(ProfileChangeLogDao.class);
+
+  private final NamedParameterJdbcTemplate jdbcTemplate;
+  private final ObjectMapper objectMapper;
+
+  public static final String CHANGELOG_TABLE = "billing_profile_changelog";
+
+  public static final String ID = "id";
+  public static final String PROFILE_ID = "profile_id";
+  public static final String CHANGES = "changes";
+  public static final String CHANGE_DATE = "change_date";
+  public static final String CHANGE_BY = "change_by";
+  public static final String CHANGE_TYPE = "change_type";
+
+  private static final List<String> COLUMNS =
+      List.of(ID, PROFILE_ID, CHANGE_DATE, CHANGE_TYPE, CHANGE_BY, CHANGES);
+  private static final String SQL_SELECT_LIST = String.join(", ", COLUMNS);
+
+  private static final String INSERT_INTO_TABLE = "INSERT INTO " + CHANGELOG_TABLE;
+  private static final String SQL_GET_BY_PROFILE_ID =
+      "SELECT %s FROM %s WHERE %s = :%s"
+          .formatted(SQL_SELECT_LIST, CHANGELOG_TABLE, PROFILE_ID, PROFILE_ID);
+
+  @Autowired
+  public ProfileChangeLogDao(NamedParameterJdbcTemplate jdbcTemplate, ObjectMapper objectMapper) {
+    this.objectMapper = objectMapper;
+    this.jdbcTemplate = jdbcTemplate;
+  }
+
+  @ReadTransaction
+  public List<ChangeLogEntry> getChangesByProfile(UUID profileId) {
+    var params = new MapSqlParameterSource().addValue(PROFILE_ID, profileId);
+    return jdbcTemplate.query(SQL_GET_BY_PROFILE_ID, params, new ChangeLogMapper(objectMapper));
+  }
+
+  @WriteTransaction
+  public Optional<UUID> recordProfileCreate(BillingProfile profile) {
+    var sql =
+        INSERT_INTO_TABLE
+            + " (profile_id, change_type, change_by, change_date) VALUES (:profile_id, :change_type, :change_by, :change_date)";
+    MapSqlParameterSource params =
+        new MapSqlParameterSource()
+            .addValue(PROFILE_ID, profile.id())
+            .addValue(CHANGE_TYPE, ChangeLogEntry.ChangeTypeEnum.CREATE.name())
+            .addValue(CHANGE_DATE, profile.createdTime().atOffset(ZoneOffset.UTC))
+            .addValue(CHANGE_BY, profile.createdBy());
+
+    var keyHolder = new DaoKeyHolder();
+    jdbcTemplate.update(sql, params, keyHolder);
+    return keyHolder.getField(ID, UUID.class);
+  }
+
+  @WriteTransaction
+  public Optional<UUID> recordProfileUpdate(UUID profileId, String user, Map<?, ?> changes) {
+    var sql =
+        INSERT_INTO_TABLE
+            + " (profile_id, change_type, change_by, changes) VALUES "
+            + "(:profile_id, :change_type, :change_by, :changes::jsonb)";
+
+    String serializedChanges = null;
+    try {
+      serializedChanges = objectMapper.writeValueAsString(changes);
+    } catch (JsonProcessingException e) {
+      logger.error(
```

Comment:
🤔 I think this is ok; if it can't serialize the changes, it logs an error and at least records an entry with a null value for the changes so there is at least some evidence in the table. I was pondering whether this should throw in such a circumstance but I don't think we want to push that burden to clients to handle and we certainly don't want a failure to log a change to fail out the parent flight.

Any idea what `changes` logs as in the log entry?

Reply:
It's just whatever the `Map.toString()` gives.

I agree that failing because of an issue with changelog serialization would be counterproductive.